core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerClientWithFileAccess.scala

```diff
@@ -0,0 +1,186 @@
+     *
+     * Does not verify that the returned file actually exists.
+     *
+     * @param containerId Id of the desired Docker container
+     * @return canonical location of the container's log file
+     */
+    protected def containerLogFile(containerId: ContainerId) =
+        new File(containerDirectory(containerId), s"${containerId.asString}-json.log").getCanonicalFile()
+
+    /**
+     * Provides the contents of the specified Docker container's configuration
+     * file as JSON object.
+     *
+     * @param configFile the container's configuration file in JSON format
+     * @return contents of configuration file as JSON object
+     */
+    protected def configFileContents(configFile: File): Future[JsObject] = Future {
+        blocking { // Needed due to synchronous file operations
+            val source = Source.fromFile(configFile)
+            val config = try source.mkString finally source.close()
+            config.parseJson.asJsObject
+        }
+    }
+
+    /**
+     * Extracts the IP of the container from the local config file of the docker daemon.
+     *
+     * A container may have more than one network. The container has an
+     * IP address in each of these networks such that the network name
+     * is needed.
+     *
+     * @param id the id of the container to get the IP address from
+     * @param network name of the network to get the IP address from
+     * @return the ip address of the container
+     */
+    protected def ipAddressFromFile(id: ContainerId, network: String): Future[ContainerIp] = {
+        configFileContents(containerConfigFile(id)).map { json =>
+            val networks = json.fields("NetworkSettings").asJsObject.fields("Networks").asJsObject
+            val specifiedNetwork = networks.fields(network).asJsObject
+            val ipAddr = specifiedNetwork.fields("IPAddress")
+            ContainerIp(ipAddr.convertTo[String])
+        }
+    }
+
+    // See implemented trait for description
+    override def inspectIPAddress(id: ContainerId, network: String)(implicit transid: TransactionId): Future[ContainerIp] = {
+        ipAddressFromFile(id, network).recoverWith {
+            case _ => super.inspectIPAddress(id, network)
+        }
+    }
+
+    // See implemented trait for description
+    def rawContainerLogs(containerId: ContainerId, fromPos: Long): Future[ByteBuffer] = Future {
+        blocking { // Needed due to synchronous file operations
+            var fis: FileInputStream = null
+            try {
+                val file = containerLogFile(containerId)
+                val size = file.length
+
+                fis = new FileInputStream(file)
+                val channel = fis.getChannel().position(fromPos)
+
+                // Buffer allocation may fail if the log file is too large to hold in memory or
+                // too few space is left on the heap, respectively.
+                var remainingBytes = (size - fromPos).toInt
```

Comment:
Note that the whole code here is copied from the old invoker code. It's true, that there may be value in refactoring it. In this specific case (log-file reading) I want to be very cautious with that though. Log files have quite a history and i'd rather not conflate the new containerpool with fixes on that front.

That's not to say we shouldn't try though. Just a disclaimer for caution.